src/services/rocksdb/backend.rs

```diff
@@ -0,0 +1,253 @@
+        ds.finish()
+    }
+}
+
+#[async_trait]
+impl kv::Adapter for Adapter {
+    fn metadata(&self) -> kv::Metadata {
+        kv::Metadata::new(
+            Scheme::Rocksdb,
+            &self.db.path().to_string_lossy(),
+            AccessorCapability::Read | AccessorCapability::Write | AccessorCapability::List,
+        )
+    }
+
+    async fn next_id(&self) -> Result<u64> {
+        let txn = self.db.transaction();
+        match txn.get("next_id").map_err(new_rocksdb_error)? {
+            Some(v) => {
+                let id = String::from_utf8_lossy(&v)
+                    .parse::<u64>()
+                    .map_err(new_rocksdb_error)?;
+                txn.put("next_id", (id + 1).to_string().as_bytes())
+                    .map_err(new_rocksdb_error)?;
+                txn.commit().map_err(new_rocksdb_error)?;
+                Ok(id + 1)
+            }
+            None => {
+                txn.put("next_id", "1".as_bytes())
+                    .map_err(new_rocksdb_error)?;
+                txn.commit().map_err(new_rocksdb_error)?;
+                Ok(1)
+            }
+        }
+    }
+
+    async fn get(&self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        self.db.get(key).map_err(new_rocksdb_error)
+    }
+
+    async fn set(&self, key: &[u8], value: &[u8]) -> Result<()> {
+        self.db.put(key, value).map_err(new_rocksdb_error)
+    }
+
+    async fn scan(&self, prefix: &[u8]) -> Result<kv::KeyStreamer> {
+        Ok(Box::new(KeyStream::new(self.db.clone(), prefix)))
+    }
+
+    async fn delete(&self, key: &[u8]) -> Result<()> {
+        self.db.delete(key).map_err(new_rocksdb_error)
+    }
+}
+
+type ScanKeysResult = Result<(Option<Vec<u8>>, Vec<Vec<u8>>)>; // (cursor, keys)
+
+#[pin_project]
+struct KeyStream {
+    db: Arc<TransactionDB>,
+    prefix: Vec<u8>,
+    cursor: Option<Vec<u8>>,
+    keys: IntoIter<Vec<u8>>,
+    fut: Option<BoxFuture<'static, ScanKeysResult>>,
+}
+
+impl KeyStream {
+    fn new(db: Arc<TransactionDB>, prefix: &[u8]) -> Self {
+        Self {
+            db,
+            prefix: prefix.to_vec(),
+            cursor: Some(prefix.to_vec()),
+            keys: vec![].into_iter(),
+            fut: None,
+        }
+    }
+}
+
+impl Stream for KeyStream {
+    type Item = Result<Vec<u8>>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+        loop {
+            if let Some(key) = this.keys.next() {
+                return Poll::Ready(Some(Ok(key)));
+            }
+
+            match this.fut {
+                None => match this.cursor.take() {
+                    None => return Poll::Ready(None),
+                    Some(cursor) => {
+                        let db = this.db.clone();
+                        let prefix = this.prefix.clone();
+                        let fut = async move {
+                            scan_keys(db.as_ref(), &prefix, &cursor, SCAN_LIMIT)
```

Comment:
This API seems a blocking function call, why we need to build a new future here?

Reply:
Ops, didn't realized it.